src/OpenTelemetry.Contrib.Instrumentation.Runtime/Implementation/EventCounterListener.cs

```diff
@@ -0,0 +1,149 @@
+﻿// <copyright file="EventCounterListener.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.Tracing;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Runtime.Implementation
+{
+    /// <summary>
+    /// EventCounterListener that subscribes to EventSource Events.
+    /// </summary>
+    internal class EventCounterListener : EventListener
+    {
+        private const string EVENTSOURCENAME = "System.Runtime";
+
+        private readonly ConcurrentQueue<EventSource> allEventSourcesCreated = new ConcurrentQueue<EventSource>();
+        private readonly Dictionary<string, string> refreshIntervalDictionary;
+        private readonly bool isInitialized = false;
+
+        private readonly IEventCounterStore eventCounterStore;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EventCounterListener"/> class.
+        /// </summary>
+        /// <param name="eventCounterStore">Instance to store the data to.</param>
+        public EventCounterListener(IEventCounterStore eventCounterStore)
+        {
+            this.eventCounterStore = eventCounterStore ?? throw new ArgumentNullException(nameof(eventCounterStore));
+
+            try
+            {
+                this.refreshIntervalDictionary = new Dictionary<string, string>();
+                this.refreshIntervalDictionary.Add("EventCounterIntervalSec", "1");
```

Comment:
This PR is dedicated to the runtime metrics and is using the API directly to retrieve the values. 
Unfortunately sometimes the API is internal, so we've to fallback to the EventCounter. 
The advantage of a this instrumentation is that it can use the direct API as soon as the runtime make it public. It will perform better as retrieving the values from event counters with less overhead. 
The EventCounter instrumentation from #177 is a general purpose implementation to retrieve any existing EventCounter.

Reply:
I think ideally we want this "Runtime Instrumentation" to use the OpenTelemetry Metrics API rather than relying on the EventCounter. The main reason is efficiency and accuracy - especially considering Prometheus scenario that [using an asynchronous instrument](https://github.com/open-telemetry/opentelemetry-dotnet/blob/f2fcb53214d737b089bad523218b1ef37dda30c2/docs/metrics/learning-more-instruments/Program.cs#L33) will make sure we only invoke the callback when Prometheus is scaping metrics, rather than having to poll the EventCounter based on some predetermined schedule.

If there are cases where we cannot use OpenTelemetry Metrics API (e.g. there is no proper API - such as ObservableUpDownCounter) or we cannot access the internal functionality of .NET runtime (e.g. if a certain EventCounter is implemented by calling some non-public API), we should discuss and see if we can find better approaches.

Reply:
> Unfortunately sometimes the API is internal, so we've to fallback to the EventCounter.

I suggest that we call out the list of APIs, and we should work with the .NET Runtime team to see if it makes sense to make these public.

Reply:
> @hananiel the goal for #207 is to have a pure OpenTelemetry Metrics API based instrumentation library for .NET runtime - that does not rely on EventCounter at all (as a consequence, provides better efficiency, sampling accuracy and probably better mode - e.g. different gen of GC could be modeled as metric dimensions/attributes rather than each modeled as a separate metric).

Reply:
@reyang These are the APIs that are internal and should be made public:
- GC.GetLastGCPercentTimeInGC()
- GC.GetGenerationSize() // for generation 0-4
- System.Reflection.Assembly.GetAssemblyCount()
- Exception.GetExceptionCount()
- RuntimeEventSourceHelper.GetCpuUsage()

Should I make PRs or are you discuss these requests internally?

Reply:
@twenzel Thank you! Let me discuss these with the .NET runtime team today, I will get back with the meeting notes.

Reply:
I'd suggest to remove any EventCounter specific code from this PR. 
in the 1st version, we can support only those metrics which are exposed by runtime. 
and as runtime exposes more, we can modify this to support them as well.

Reply:
Ok, EventCounter specific code was removed.

Reply:
> @reyang These are the APIs that are internal and should be made public:
> 
> * GC.GetLastGCPercentTimeInGC()
> * GC.GetGenerationSize() // for generation 0-4
> * System.Reflection.Assembly.GetAssemblyCount()
> * Exception.GetExceptionCount()
> * RuntimeEventSourceHelper.GetCpuUsage()
> 
> Should I make PRs or are you discuss these requests internally?

@cijothomas and I had a discussion with the .NET runtime team (@maryamariyan, @noahfalk and @tarekgh), here goes a quick summary:

1. Some of the methods (e.g. `RuntimeEventSourceHelper.GetCpuUsage()`) were introduced into the runtime to reduce dependencies, there is likely going to be public APIs in other .NET runtime assemblies that meet the requirements here.
2. Some methods might have no good alternatives that are public. From .NET perspective, @noahfalk can help to see what's the direction - e.g. we might make certain API public in the future version of runtime, or design a public APIs that provide the same or even more feature.
3. Given this OpenTelemetry package will have to work with existing versions of the runtime, we need use reflection (probably in combination with some conditional compilation or runtime detection), for the functions that we took reflection approach here, .NET team could add comment in the source code to avoid breaking OpenTelemetry accidentally - before we can fully move to public APIs.
4. All of the above doesn't need to block this PR, we can address these 5 missing instruments in separate PRs.

Reply:
@reyang @cijothomas - Talking with Maoni about the percent-time-in-gc counter a similar issue arises that I see with the cpu % counter. It seems like what we'd want to do is represent these things as ObservableCounters which measure the number of seconds where something was true:
CPU = seconds/interval where CPU was in use
GC = seconds/interval where GC had all threads paused

Reply:
Take an example to make sure I understand the proposal:

| Time since program started | Total time when GC had all threads paused |
| -------------------------- | ----------------------------------------- |
| 1.5 seconds                | 0.1 second                                |
| 6.5 seconds                | 1.1 second                                |
| 16.5 seconds               | 2.1 second                                |

* T0: 0s
* T1: 1.5s
* T2: 6.5s
* T3: 16.5s

Option 1:

* ObservableCounter("process.runtime.dotnet.gc.time")
* at T1, record 0.1 // (T0, T1)
* at T2, record 1.1 // (T0, T2)
* at T3, record 2.1 // (T0, T3)

The consumer can configure the exporter to use delta temporality, so proper rate can be calculated downstream.

Option 2:

* at T1, record 0.067 // 0.1 / 1.5
* at T2, record 0.2 // (1.1 - 0.1) / (6.5 - 1.5)
* at T3, record 0.1 // (2.1 - 1.1) / (16.5 - 6.5)

It seems Option 2 won't work, as the data being recorded won't fit any existing instrument.

Reply:
Yep exactly, option 1 is what I was suggesting we should start doing and option 2 is the problematic path that the cpu-usage EventCounter is doing right now.

Option (2) works ok with an ObservableGauge only if there is guaranteed exactly one observer/exporter, allowing the callback to infer the interval by timestamping when it was last called. As soon as there is a 2nd observer then the callback can't distinguish if the call at T2 is a new observer trying to measure the interval [0, T2] or the same observer that called at T1 now trying to measure [T1, T2].

At risk of getting a little into the weeds, the current runtime implementation of GetLastGCPercentTimeInGC() doesn't capture a timestamp of when it was last called, but the way it avoids doing that produces a measurement people are likely to find confusing. In your example above, imagine that the various GCs occured at:
GC 1: 1.0-1.2  (0.1 sec paused)
GC 2: 2.5-3.9  (0.9 sec paused)
GC 3: 4.0-4.1  (0.1 sec paused)
GC 4: 15.0-16.2 (1 sec paused)
You can see GCs can also do some work while the threads aren't paused. Adding up the pause times gives the same 0.1, 1.1 and 2.1 measurements as above when observed at T=1.5, 6.5, and 16.5.
If you measure at some time T, GetLastGCPercentTimeInGC() finds the most recent GC X that finished before time T and calculates GC_x_pause/ ( GC_x_end - GC_x-1_end). For example if you query at T=6.5 it determines GC 3 was the most recent completed GC before that point which ended at 4.1. Before that GC 2 ended at 3.9. That gives 0.1 pause time/(4.1-3.9) = 50% time in GC. Users were probably hoping to see (1.1-0.1)/(6.5-1.5)=20% time in GC and will be surprised not to see that. So the current GC stats don't suffer non-determinism when subjected to multiple observers, but only by using a surprising calculation that isn't necessarily a good approximation of what users really wanted.